### PR TITLE
[#1568] Add restricted API to allow access to internals of the node model

### LINKED ISSUE #1568: Make basic* methods of AbstractNode accessible to clients
AbstractNode has a few methods that offer more efficient traversal options for a small set of use cases. These are not accessible to clients without resorting to reflection.

Add static utils to InternalNodeModelUtils to allow clients to benefit from these at their own risk.

### PR DESCRIPTION
**Pull request by szarnekow**
closes #1568

### CODE REVIEW

File: org.eclipse.xtext/src/org/eclipse/xtext/nodemodel/impl/InternalNodeModelUtils.java [hunk 1]
```diff
@@ -51,6 +56,36 @@ protected static LineAndColumn getLineAndColumn(INode anyNode, int documentOffse
 		return getLineAndColumn(document, lineBreaks, documentOffset);
 	}
 	
+	/**
+	 * Obtain the parent node of the given node. Does not unfold {@link SyntheticCompositeNode synthetic}
+	 * nodes. May return null.
```
Comment (kthoms) [inline org.eclipse.xtext/src/org/eclipse/xtext/nodemodel/impl/InternalNodeModelUtils.java:61]:
```suggestion
	 * nodes. May return <code>null</code>.
```

File: org.eclipse.xtext/src/org/eclipse/xtext/nodemodel/impl/InternalNodeModelUtils.java [hunk 2]
```diff
@@ -51,6 +56,36 @@ protected static LineAndColumn getLineAndColumn(INode anyNode, int documentOffse
...
+	 * 
+	 * @since 2.24
+	 */
+	protected static CompositeNode basicGetParent(AbstractNode node) {
+		return node.basicGetParent();
+	}
+	
+	/**
+	 * Obtain the first child node of the given node. Does not unfold {@link SyntheticCompositeNode synthetic}
+	 * nodes. May return null.
```
Comment (kthoms) [inline org.eclipse.xtext/src/org/eclipse/xtext/nodemodel/impl/InternalNodeModelUtils.java:71]:
```suggestion
	 * nodes. May return <code>null</code>.
```

File: org.eclipse.xtext/src/org/eclipse/xtext/nodemodel/impl/InternalNodeModelUtils.java [hunk 3]
```diff
@@ -51,6 +56,36 @@ protected static LineAndColumn getLineAndColumn(INode anyNode, int documentOffse
...
+	 * 
+	 * @since 2.24
+	 */
+	protected static AbstractNode basicGetFirstChild(CompositeNode node) {
+		return node.basicGetFirstChild();
+	}
+	
+	/**
+	 * Obtain the last child node of the given node. Does not unfold {@link SyntheticCompositeNode synthetic}
+	 * nodes. May return null.
```
Comment (kthoms) [inline org.eclipse.xtext/src/org/eclipse/xtext/nodemodel/impl/InternalNodeModelUtils.java:81]:
```suggestion
	 * nodes. May return <code>null</code>.
```

**Review by tivervac — APPROVED** [review]
LGTM

File: org.eclipse.xtext/src/org/eclipse/xtext/nodemodel/impl/InternalNodeModelUtils.java [hunk 4]
```diff
@@ -19,7 +19,7 @@
 import com.google.common.collect.Lists;
```
Comment (kthoms) [inline org.eclipse.xtext/src/org/eclipse/xtext/nodemodel/impl/InternalNodeModelUtils.java:20]: Update copyright year (I could not add the comment to the right line)

Comment (szarnekow) [reply org.eclipse.xtext/src/org/eclipse/xtext/nodemodel/impl/InternalNodeModelUtils.java:20]: Done

File: org.eclipse.xtext/src/org/eclipse/xtext/nodemodel/impl/InternalNodeModelUtils.java (hunk 1 above)
Comment (szarnekow) [reply org.eclipse.xtext/src/org/eclipse/xtext/nodemodel/impl/InternalNodeModelUtils.java:61]: I'll use `{@code null}`

Comment (kthoms) [reply org.eclipse.xtext/src/org/eclipse/xtext/nodemodel/impl/InternalNodeModelUtils.java]: even better

**Review by kthoms — APPROVED** [review]

**Pull request merged**